### DISCUSSION TITLE
CodeQL adds beta support for Kotlin! 🚀

### DISCUSSION THREAD

**Discussion by turbo** [github/codeql#11460] [Show and tell]
Starting today, [GitHub code scanning](https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/about-code-scanning) includes beta support for analyzing code written in Kotlin, powered by the CodeQL engine.

Kotlin is a key programming language used in the creation of Android mobile applications, and is an increasingly popular choice for new projects, augmenting or even replacing Java. To help organisations and open source developers find potential vulnerabilities in their code, we’ve added Kotlin support (beta) to the CodeQL engine that powers [GitHub code scanning](https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/about-code-scanning). CodeQL now natively supports Kotlin, as well as mixed Java and Kotlin projects. [Set up code scanning](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning-with-codeql) on your repositories today to receive actionable [security alerts right on your pull-requests](https://github.blog/changelog/2022-06-02-users-can-view-and-comment-on-code-scanning-alerts-on-the-conversation-tab-in-a-pull-request/). To enable Kotlin analysis on a repository, [configure the code scanning workflow languages](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-the-codeql-workflow-for-compiled-languages) to include java.

CodeQL support for Kotlin has already been used to identify novel real-world vulnerabilities in popular apps, from [task management](https://github.com/tasks/tasks/security/advisories/GHSA-8x58-cg74-8jg8) to [productivity platforms](https://nvd.nist.gov/vuln/detail/CVE-2021-43863). You can watch the GitHub Universe talk on how CodeQL was used to identify vulnerabilities like these [here](https://www.youtube.com/watch?v=P1wqo276KjU).

Kotlin beta support is available by default in [GitHub.com](http://github.com/) code scanning, the [CodeQL CLI](https://github.com/github/codeql-cli-binaries/releases), and the [CodeQL extension for VS Code](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-codeql). GitHub Enterprise Server (GHES) version 3.8 will include this beta release.

If you have any feedback or questions, please use this discussion thread or open an issue if you encounter any problems.

**Comment by sschuberth** [comment]
I tried to enable Kotlin analysis in [here](https://github.com/oss-review-toolkit/ort/pull/6134), but the build after initializing CodeQL is [running into `java.lang.OutOfMemoryError: Java heap space`](https://github.com/oss-review-toolkit/ort/actions/runs/3568240298/jobs/5996909792) although a regular build just works fine.

@turbo, can it really be that a build after CodeQL initialization takes more memory?

PS: I tried with *autobuild* first, but changed to a manual build step after I saw the same memory issue originally there.

**Reply by turbo** [reply]
Hi @sschuberth,

I've raised this issue internally. The team is investigating and will get back to you with next steps soon.

**Reply by tamasvajk** [reply]
Hello @sschuberth, 

I could reproduce your `java.lang.OutOfMemoryError: Java heap space` issue locally with the following build command `./gradlew clean classes -x :reporter-web-app:yarnBuild`. To some extent, it's expected that the analysis takes more memory than a normal build. We're going to investigate if we can improve on our memory footprint. For the time being, I was able to overcome the issue by adding `-Dorg.gradle.jvmargs=-Xmx1g` to the build command. 

Let us know if this doesn't unblock you.

**Comment by LeShadow** [comment]
Hey @turbo ,

I've been following this closely because this was a requested language for us!

Am I correct to assume that in your CodeQL file you have to add both `java` AND `kotlin` to the language matrix?

Thank you!

**Reply by turbo** [reply]
Hi @LeShadow,

Good question! No, Kotlin is part of our Java support. You only have to enable `java`, which from now on automatically includes Kotlin (even for Kotlin-only projects).

**Reply by LeShadow** [reply]
@turbo thank you for the clarification! 

I did a test run on one of our repositories, and either we have super good code quality or I am missing something. Because I have no alerts whatsoever on an entire codebase. Is this possible with the default settings?

**Reply by turbo** [reply]
@LeShadow To check if CodeQL has successfully analyzed Kotlin code, take a look at the Actions logs for the workflow. It should include a section on "analyzed LoC" (lines of code). If those roughly match the expected count, the analysis worked as intended.

If that's the case you can then take a look at increasing the number of queries run by, for example, [enabling the `security-extended` query suite](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs). 

A note on:

> super good code quality

Just to avoid confusion: CodeQL's focus is code security, not code quality. So while extended the number of queries run may flag up additional security alerts, it won't include quality checks. For that use case, we recommend running additional quality tools that integrate with GitHub code scanning, such as [Detekt](https://detekt.dev/docs/introduction/reporting/#integration-with-github-code-scanning).

If you still encounter problems after following these steps, please open an issue and the team will assist 🙂.

**Reply by sschuberth** [reply]
> No, Kotlin is part of our Java support. You only have to enable `java`, which from now on automatically includes Kotlin (even for Kotlin-only projects).

How about simply defining "kotlin" as an alias for "java" to avoid some potential user confusion? Also, if users would start using "kotlin" already now, that would allow you guys to seamlessly introduce Kotlin-specific checks in the future.

**Reply by turbo** [reply]
@sschuberth Thanks for the feedback. The mapping is automatically done when code scanning is set up via the UI for the first time on a repo containing Kotlin and we're considering various ways of making this more seamless for our integrated languages (Java/Kotlin, JS/TS).

**Comment by sschuberth** [comment]
So, I've run into a new problem now. After [increasing the heap memory](https://github.com/github/codeql/discussions/11460#discussioncomment-4286323) helped to complete a CodeQL scan, three `Arbitrary file write during archive extraction ("Zip Slip")` issues were found, which is great. However, these [findings stick around](https://github.com/oss-review-toolkit/ort/pull/6134/checks?check_run_id=9841546841) even after [addressing them](https://github.com/oss-review-toolkit/ort/pull/6134/commits/9ea2ccead5fd52daa1265edbf3bf73f7c6ebdad7).

Can it be that CodeQL cannot deal with Kotlin's `File.startWith()` extension function to recognize the check, @tamasvajk?

**Reply by tamasvajk** [reply]
Thanks for reporting this. We're fixing the query.

**Reply by sschuberth** [reply]
Please tell me when it's worth to rerun the check to verify the fix. Thanks in advance!

**Reply by tamasvajk** [reply]
This is the PR fixing the issue: https://github.com/github/codeql/pull/11550.

**Reply by sschuberth** [reply]
Is this already expected to be in a release? Because I just rebased https://github.com/oss-review-toolkit/ort/pull/6134, but it still shows the fixed violations.

**Reply by smowton** [reply]
No; the last three releases have been from a stable branch and so haven't included this change. It is expected to appear mid-January and will include that PR.

**Reply by sschuberth** [reply]
I was just [trying again](https://github.com/oss-review-toolkit/ort/actions/runs/3965751481/jobs/6795807240), as the [0.5.0 release notes](https://github.com/github/codeql/blob/codeql-cli/v2.12.0/java/ql/lib/CHANGELOG.md) sound as if this fix was included. But now I'm running into the problem that Kotlin 1.8 is not supported yet. Any plans to support that soon?

**Reply by atorralba** [reply]
Support for Kotlin 1.8 has been [recently merged](https://github.com/github/codeql/pull/11637) — it should be available in the default GitHub Actions image by 15th February.

**Reply by smowton** [reply]
You can also experimentally opt into the latest release by using the `tools: latest` parameter to the `github/codeql-action/init` action. For example:

```
    - name: Initialize CodeQL
      uses: github/codeql-action/init@v2
      with:
        tools: latest
```

Note this slows the action down due to downloading the CodeQL bundle, so be sure to only use this if necessary, and schedule switching it off when no longer required.

**Reply by sschuberth** [reply]
Great, thanks for the hint, I'll give it a try!

**Reply by smowton** [reply]
Sorry I should clarify, it hasn't been released yet but will be in the next release, expected around January 25th. This flag will allow you to try out the latest release at that time rather than wait until mid-February.

**Reply by sschuberth** [reply]
Ah, right, that's why [I still see it failing](https://github.com/oss-review-toolkit/ort/actions/runs/3967309451/jobs/6799138281#step:5:202) even with `tools: latest` currently. I'll try again after January 25th.

**Reply by sschuberth** [reply]
So, it was working for a short while with Kotlin 1.8, but now [breaks again with Kotlin 1.8.10](https://github.com/oss-review-toolkit/ort/actions/runs/4101266246/jobs/7072895335#step:5:115) 😭 Is there are way to not break on patch-level-only Kotlin updates?

**Reply by sschuberth** [reply]
Looks like support for Kotlin 1.8.10 (and 1.8.20) [was added](https://github.com/github/codeql/pull/12090), but [not released yet](https://github.com/github/codeql/pull/12173#issuecomment-1428455090).

**Reply by marcphilipp** [reply]
@smowton Can you provide an estimate when support for 1.8.10 will be released? Gradle 8.0 was released recently and uses 1.8.10 so currently all builds on that Gradle version fail.

**Reply by smowton** [reply]
A release cycle just began that includes the required support, so this should be available with `tools: latest` from roughly Tuesday if the release goes smoothly.

**Reply by marcphilipp** [reply]
@smowton Thanks for your reply! I tried finding docs on `tools: latest` without any luck. Is that GitHub Actions workflow syntax?

**Reply by atorralba** [reply]
@marcphilipp It's explained in a comment above: https://github.com/github/codeql/discussions/11460#discussioncomment-4736473.

**Reply by sschuberth** [reply]
> A release cycle just began that includes the required support, so this should be available with `tools: latest` from roughly Tuesday if the release goes smoothly.

Is there any GitHub project we can watch for releases to get notified when Kotlin 1.8.10 support should be available?

**Reply by marcphilipp** [reply]
I'd also like that since I keep manually rerunning the job to see if it will eventually succeed but it's still red as of just now.

**Reply by smowton** [reply]
https://github.com/github/codeql-cli-binaries/releases

Since that has been released, with `tools: latest` you should see CodeQL 2.12.3 in use and Kotlin 1.8.10 supported.

**Reply by sschuberth** [reply]
@smowton I'm still seeing [this](https://github.com/oss-review-toolkit/ort/actions/runs/4261085944/jobs/7415011229#step:5:151) with [tools: latest](https://github.com/oss-review-toolkit/ort/blob/605dedbcf88da096f33fc7e42f2cdd63caacb12e/.github/workflows/build-and-test.yml#L60) in a [PR](https://github.com/oss-review-toolkit/ort/pull/6548) that I've created about 1 hour ago.

**Reply by smowton** [reply]
Apologies, I missed that in fact we need both a CLI binary (released yesterday) and a corresponding bump to https://github.com/github/codeql-action/tree/releases/v2, which typically follows around a day or two later.

**Reply by smowton** [reply]
Edit: said bump has now been applied. Rerunning your workflow ought to work.

**Reply by sschuberth** [reply]
Confirmed, thanks for the heads-up!

**Comment by eygraber** [comment]
I integrated CodeQL into my [Kotlin project](https://github.com/eygraber/portal/blob/master/.github/workflows/codeql.yml) but the action log and CSV show that Kotlin files weren't extracted. Is there something that needs to be done to let CodeQL know that there are Kotlin files present?

**Reply by sschuberth** [reply]
What looks a bit weird is that you set up `gradle-build-action`, but when use a manual `run` step to build (instead of passing `arguments` to `gradle-build-action`). Though I don't think that should be the root cause of things not working in your case. Anyway, [this](https://github.com/oss-review-toolkit/ort/blob/707023526d7c08c899e18ca3a25ab39af41b6507/.github/workflows/build-and-test.yml#L50-L74) works for us.

**Reply by eygraber** [reply]
I have another repo where that works fine. With this one I'm noticing a warning "Timed out waiting for analysis to finish processing"

**Reply by sschuberth** [reply]
Probably a memory issue. Try increasing the memory for building similar to like we did.

**Reply by Marcono1234** [reply]
@eygraber, were you able to solve this? I assume you fixed it by disabling the build cache in https://github.com/eygraber/portal/commit/182ac6e9cbf82a9d89375d339368a0fbb98f3f47, right?

**Reply by eygraber** [reply]
Yes, sorry I forgot I had two places where I was discussing this. Disabling the caching solved the problem.

**Comment by NorseDreki** [comment]
Hi there,

Looks like support for Kotlin actually means support for Kotlin code only targeting JVM, is that assumption correct?
From what I see, is there currently no support for Kotlin / Native code which compiles to platform native binary?

**Reply by smowton** [reply]
That's right. That could change in the future, but right now our support works as part of a combined Java/Kotlin JVM languages analysis.

**Reply by NorseDreki** [reply]
Thanks for heads up.
